### PR TITLE
Improve robustness of test_exit_code_127 k8s test

### PR DESCRIPTION
**Pull request by mvdbeek**
Maybe fixes
https://github.com/galaxyproject/galaxy/runs/6941388460?check_suite_focus=true#step:14:909,
but seems like a good change on its own.

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these contributions under [Galaxy's current license](https://github.com/galaxyproject/galaxy/blob/dev/LICENSE.txt).
- [x] I agree to allow the Galaxy committers to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT). If this condition is an issue, uncheck and just let us know why with an e-mail to galaxy-committers@lists.galaxyproject.org.

### CODE REVIEW

File: test/integration/test_kubernetes_runner.py [hunk 1]
```diff
@@ -308,18 +308,20 @@ def test_external_job_delete(self):
...
         app = self._app
-        sa_session = app.model.context.current
-        job = sa_session.query(app.model.Job).get(app.security.decode_id(running_response.json()["jobs"][0]["id"]))
+        job_id = app.security.decode_id(running_response.json()["jobs"][0]["id"])
+        sa_session = app.model.context
+        job = sa_session.query(app.model.Job).get(job_id)
+        self._wait_for_external_state(sa_session=sa_session, job=job, expected=app.model.Job.states.RUNNING)
+
         external_id = job.job_runner_external_id
         output = unicodify(subprocess.check_output(["kubectl", "logs", "-l" f"job-name={external_id}"]))
```
Comment (mvdbeek) [inline test/integration/test_kubernetes_runner.py]:
```suggestion
        output = unicodify(subprocess.check_output(["kubectl", "logs", "-l", f"job-name={external_id}"]))
```

File: test/integration/test_kubernetes_runner.py [hunk 2]
```diff
@@ -308,20 +309,28 @@ def test_external_job_delete(self):
...
-        job = sa_session.query(app.model.Job).get(app.security.decode_id(running_response.json()["jobs"][0]["id"]))
+        job_id = app.security.decode_id(running_response.json()["jobs"][0]["id"])
+        sa_session = app.model.context
+        job = sa_session.query(app.model.Job).get(job_id)
+        self._wait_for_external_state(sa_session=sa_session, job=job, expected=app.model.Job.states.RUNNING)
+
         external_id = job.job_runner_external_id
-        output = unicodify(subprocess.check_output(["kubectl", "logs", "-l" f"job-name={external_id}"]))
+        log_cmd = ["kubectl", "logs", "-l", f"job-name={external_id}"]
+        p = subprocess.run(log_cmd)
```
Comment (nsoranzo) [inline test/integration/test_kubernetes_runner.py]:
```suggestion
        p = subprocess.run(log_cmd, capture_output=True, text=True)
```

File: test/integration/test_kubernetes_runner.py [hunk 3]
```diff
@@ -308,20 +308,28 @@ def test_external_job_delete(self):
...
+        job = sa_session.query(app.model.Job).get(job_id)
+        self._wait_for_external_state(sa_session=sa_session, job=job, expected=app.model.Job.states.RUNNING)
+
         external_id = job.job_runner_external_id
-        output = unicodify(subprocess.check_output(["kubectl", "logs", "-l" f"job-name={external_id}"]))
+        log_cmd = ["kubectl", "logs", "-l", f"job-name={external_id}"]
+        p = subprocess.run(log_cmd, capture_output=True, text=True)
+        if p.returncode:
+            raise Exception(
+                f"Command '{' '.join(log_cmd)}' failed with exit code: {p.returncode}.\nstdout: {p.stdout}\nstderr: {p.stderr}"
```
Comment (nsoranzo) [inline test/integration/test_kubernetes_runner.py]: What about adding the following to `lib/galaxy/util/__init__.py`:

```python
import shlex

try:
    shlex_join = shlex.join
except AttributeError:
    # Python < 3.8
    shlex_join = lambda split_command: " ".join(map(shlex.quote, split_command))
```

and then:

```suggestion
                f"Command '{shlex_join(log_cmd)}' failed with exit code: {p.returncode}.\nstdout: {p.stdout}\nstderr: {p.stderr}"
```

We can reuse such `shlex_join()` also in `lib/galaxy/datatypes/text.py`:222 , `lib/galaxy/tool_util/deps/conda_util.py`:240, `lib/galaxy/tool_util/deps/mulled/mulled_build.py`:304 and `lib/galaxy/util/path/__init__.py`:368 .

**Comment by mvdbeek** [comment]
Seems to have worked at least once now 🤞

File: test/integration/test_kubernetes_runner.py [hunk 4]
```diff
@@ -308,20 +312,36 @@ def test_external_job_delete(self):
...
-        output = unicodify(subprocess.check_output(["kubectl", "logs", "-l" f"job-name={external_id}"]))
+
+        def get_kubectl_logs(allow_wait=True):
+            log_cmd = ["kubectl", "logs", "-l", f"job-name={external_id}"]
+            p = subprocess.run(log_cmd, capture_output=True, text=True)
+            if p.returncode:
+                if allow_wait and "is waiting to start" in p.stderr:
+                    return None
+                raise Exception(
+                    f"Command '{shlex_join}' failed with exit code: {p.returncode}.\nstdout: {p.stdout}\nstderr: {p.stderr}"
```
Comment (nsoranzo) [inline test/integration/test_kubernetes_runner.py]:
```suggestion
                    f"Command '{shlex_join(log_cmd)}' failed with exit code: {p.returncode}.\nstdout: {p.stdout}\nstderr: {p.stderr}"
```

File: lib/galaxy/util/__init__.py [hunk 5]
```diff
@@ -77,6 +78,12 @@
     safe_relpath,
 )
 
+try:
+    shlex_join = shlex.join  # type: ignore[attr-defined]
+except AttributeError:
+    # Python < 3.8
+    shlex_join = lambda split_command: " ".join(map(shlex.quote, split_command))  # noqa: E731
```
Comment (nsoranzo) [inline lib/galaxy/util/__init__.py]: I forgot about E731, I am also fine with:
```suggestion
    def shlex_join(split_command):
        return " ".join(map(shlex.quote, split_command))
```

**Review by nsoranzo — APPROVED** [review]

**Pull request merged**